### PR TITLE
Add support for search model api

### PR DESCRIPTION
**Pull request by VijayanB**
### Description
Allow users to search model similar to search requests. 
Example:
Using GET
```
GET /_plugins/_knn/models/_search?<query_filters>
{
    "query": {
         ...
     }
}

{
  "took": 13,
  "timed_out": false,
  "_shards": {
    "total": 5,
    "successful": 5,
    "skipped": 0,
    "failed": 0
  },
  "hits": {
    ...
  }
} 

```
Using POST
```
POST /_plugins/_knn/models/_search?<query_filters>
{
    "query": {
         ...
     }
}

{
  "took": 13,
  "timed_out": false,
  "_shards": {
    "total": 5,
    "successful": 5,
    "skipped": 0,
    "failed": 0
  },
  "hits": {
    ...
  }
} 
```
 
### Issues Resolved
[List any issues this PR will resolve]
 
### Check List
- [x] New functionality includes testing.
  - [x] All tests pass
- [ ] New functionality has been documented.
  - [ ] New functionality has javadoc added
- [x] Commits are signed as per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/k-NN/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: src/main/java/org/opensearch/knn/plugin/rest/RestSearchModelHandler.java [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+                    RestRequest.Method.POST,
+                    String.format(Locale.ROOT, "%s/%s/%s", KNNPlugin.KNN_BASE_URI, MODELS, SEARCH)
+                )
+            );
+    }
+
+    @Override
+    protected RestChannelConsumer prepareRequest(RestRequest request, NodeClient client) throws IOException {
+        SearchSourceBuilder searchSourceBuilder = new SearchSourceBuilder();
+        searchSourceBuilder.parseXContent(request.contentOrSourceParamParser());
```
Comment (jmazanec15) [inline src/main/java/org/opensearch/knn/plugin/rest/RestSearchModelHandler.java:74]: What happens when user passes query params? i.e. source_includes/source_excludes?

File: src/main/java/org/opensearch/knn/plugin/rest/RestSearchModelHandler.java [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+
+        SearchRequest searchRequest = new SearchRequest().source(searchSourceBuilder).indices(MODEL_INDEX_NAME);
+        return channel -> client.execute(SearchModelAction.INSTANCE, searchRequest, search(channel));
+    }
+
+    private RestResponseListener<SearchResponse> search(RestChannel channel) {
```
Comment (jmazanec15) [inline src/main/java/org/opensearch/knn/plugin/rest/RestSearchModelHandler.java:80]: Why is this needed? Why cant you just do as they do in search: https://github.com/opensearch-project/OpenSearch/blob/main/server/src/main/java/org/opensearch/rest/action/search/RestSearchAction.java#L134?

File: src/main/java/org/opensearch/knn/plugin/rest/RestSearchModelHandler.java [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+                )
+            );
+    }
+
+    @Override
+    protected RestChannelConsumer prepareRequest(RestRequest request, NodeClient client) throws IOException {
+        SearchSourceBuilder searchSourceBuilder = new SearchSourceBuilder();
+        searchSourceBuilder.parseXContent(request.contentOrSourceParamParser());
+
+        SearchRequest searchRequest = new SearchRequest().source(searchSourceBuilder).indices(MODEL_INDEX_NAME);
```
Comment (jmazanec15) [inline src/main/java/org/opensearch/knn/plugin/rest/RestSearchModelHandler.java:76]: Philosophical question: Should indices be set in modelDao to hide details about the index?

File: src/main/java/org/opensearch/knn/indices/Model.java [hunk 4]
```diff
@@ -69,9 +75,9 @@ public Model(ModelMetadata modelMetadata, @Nullable byte[] modelBlob) {
      * @param modelBlob binary representation of model template index. Can be null if model is not yet in CREATED state.
      * @param modelID model identifier
      */
-    public Model(ModelMetadata modelMetadata, @Nullable byte[] modelBlob, String modelID) {
-        this(modelMetadata,modelBlob);
-        this.modelID = Objects.requireNonNull(modelID, "model id must not be null");
+    public Model(ModelMetadata modelMetadata, @Nullable byte[] modelBlob, @NonNull String modelID) {
```
Comment (jmazanec15) [inline src/main/java/org/opensearch/knn/indices/Model.java:78]: NonNull doesnt provide runtime validation does it? Why remove requireNonNull?

Comment (VijayanB) [reply src/main/java/org/opensearch/knn/indices/Model.java:78]: Ack

File: src/main/java/org/opensearch/knn/plugin/rest/RestSearchModelHandler.java (hunk 2 above)
Comment (jmazanec15) [reply src/main/java/org/opensearch/knn/plugin/rest/RestSearchModelHandler.java:107]: ah I see makes sense.

File: src/main/java/org/opensearch/knn/plugin/rest/RestSearchModelHandler.java (hunk 1 above)
Comment (jmazanec15) [reply src/main/java/org/opensearch/knn/plugin/rest/RestSearchModelHandler.java]: Makes sense

**Review by jmazanec15 — APPROVED** [review]
LGTM
